processor/src/main/java/org/mapstruct/ap/internal/model/source/SubClassMappingOptions.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.internal.model.source;
+
+import java.util.Set;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.VariableElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+
+import org.mapstruct.ap.internal.gem.SubClassMappingGem;
+import org.mapstruct.ap.internal.gem.SubClassMappingsGem;
+import org.mapstruct.ap.internal.util.FormattingMessager;
+import org.mapstruct.ap.internal.util.TypeUtils;
+
+import static org.mapstruct.ap.internal.util.Message.SUBCLASSMAPPING_ILLEGAL_SUBCLASS;
+import static org.mapstruct.ap.internal.util.Message.SUBCLASSMAPPING_METHOD_SIGNATURE_NOT_SUPPORTED;
+import static org.mapstruct.ap.internal.util.Message.SUBCLASSMAPPING_NO_VALID_SUPERCLASS;
+
+/**
+ * Represents a sub class mapping as configured via {@code @SubClassMapping}.
+ *
+ * @author Ben Zegveld
+ */
+public class SubClassMappingOptions extends DelegatingOptions {
+
+    private TypeMirror sourceClass;
+    private TypeMirror targetClass;
+
+    public SubClassMappingOptions(TypeMirror sourceClass, TypeMirror targetClass, DelegatingOptions next) {
+        super( next );
+        this.sourceClass = sourceClass;
+        this.targetClass = targetClass;
+    }
+
+    @Override
+    public boolean hasAnnotation() {
+        return sourceClass != null && targetClass != null;
+    }
+
+    private static boolean isConsistent(SubClassMappingGem gem, ExecutableElement method, FormattingMessager messager,
+                                        TypeUtils typeUtils) {
+
+        if ( method.getReturnType().getKind() == TypeKind.VOID ) {
+            messager.printMessage( method, gem.mirror(), SUBCLASSMAPPING_METHOD_SIGNATURE_NOT_SUPPORTED );
+            return false;
+        }
+
+        TypeMirror sourceSubClass = gem.source().getValue();
+        TypeMirror targetSubClass = gem.target().getValue();
+        TypeMirror targetParentType = method.getReturnType();
+        boolean isConsistent = true;
+
+        boolean isChildOfAParameter = false;
+        for ( VariableElement parameter : method.getParameters() ) {
```

Comment:
Doing `method.getParameters()` is not quite correct since this will also take `@Context` parameters into consideration.  You can pass the real source parameters from the `MethodRetrievalProcessor`, since we have them there. I think you can use them there. There is also `targetParameter` and the `resultType` there as well.